From 49a772e7d7b67b006b3d18eec7b68cf98178e451 Mon Sep 17 00:00:00 2001
From: "R, Kannappan" <r.kannappan@intel.com>
Date: Wed, 10 Dec 2014 14:11:55 +0530
Subject: [PATCH] intel_soc_pmic: Fix the interrupt issue while waking from S3

While waking from S3, pmic driver tries to read the interrupt status register
using I2C call. But the I2C master did not resume. Since CRC interrupt is
configured as EDGE triggered, the system did not able to complete this
interrupt.  Hence the VBUS notification is not sent to USB driver. Also further
PMIC interrupts are disabled the interrupt is not cleared.

Have the CRC interrupt as LEVEL triggered, which shall retry the interrupt.
Also during interrupt processing, do not call the nested handlers when read/write
PMIC I2C calls failed.

Change-Id: If53f4c3ffc3ebc0fe630ed7ce4d5c35e579805a9
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-9711
Signed-off-by: R, Kannappan <r.kannappan@intel.com>
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
Reviewed-on: https://android.intel.com:443/310067
---
 drivers/mfd/intel_soc_pmic_core.c | 2 +-
 drivers/mfd/intel_soc_pmic_crc.c  | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/mfd/intel_soc_pmic_core.c b/drivers/mfd/intel_soc_pmic_core.c
index 86d8d84..63198a0 100644
--- a/drivers/mfd/intel_soc_pmic_core.c
+++ b/drivers/mfd/intel_soc_pmic_core.c
@@ -347,7 +347,7 @@ static irqreturn_t pmic_irq_thread(int irq, void *data)
 		if (test_bit(i % 32, &(pmic->irq_mask[i / 32])))
 			continue;
 
-		if (pmic_regmap_read(&pmic->irq_regmap[i].status)) {
+		if (pmic_regmap_read(&pmic->irq_regmap[i].status) > 0) {
 			pmic_regmap_write(&pmic->irq_regmap[i].ack,
 				pmic->irq_regmap[i].ack.mask);
 			handle_nested_irq(pmic->irq_base + i);
diff --git a/drivers/mfd/intel_soc_pmic_crc.c b/drivers/mfd/intel_soc_pmic_crc.c
index 47b7a2b..6c09518 100644
--- a/drivers/mfd/intel_soc_pmic_crc.c
+++ b/drivers/mfd/intel_soc_pmic_crc.c
@@ -177,7 +177,7 @@ static int crystal_cove_init(void)
 
 struct intel_soc_pmic crystal_cove_pmic = {
 	.label		= "crystal cove",
-	.irq_flags	= IRQF_TRIGGER_RISING | IRQF_ONESHOT,
+	.irq_flags	= IRQF_TRIGGER_HIGH | IRQF_ONESHOT,
 	.init		= crystal_cove_init,
 	.cell_dev	= crystal_cove_dev,
 	.irq_regmap	= crystal_cove_irqregmap,
-- 
1.9.1

